src/gurt/telemetry.c

```diff
@@ -0,0 +1,1864 @@
+ * accessed directly.  Otherwise, a lookup is performed on the path name
+ * provided in order to find the specified item.  If the item cannot be found,
+ * it is created.  If the item is created, the callers pointer is initialized
+ * for this item.  The pointer is used for direct access on subsequent calls for
+ * faster access.
+ */
+int
+d_tm_decrement_gauge(d_tm_node_t **metric, uint64_t value, char *item, ...)
+{
+	d_tm_node_t *node = NULL;
+	char path[D_TM_MAX_NAME_LEN] = {};
+	char *str;
+	int rc = D_TM_SUCCESS;
+
+	if (metric && *metric)
+		node = *metric;
+	else {
+		va_list args;
+		va_start(args, item);
+		sprintf(path, "%s", item);
+		str = va_arg(args, char *);
+		while (str) {
+			/*
+			 * verify that adding the next str token + '/'
+			 * will fit into the path string buffer
+			 * If it fits, append it to the path
+			 */
+			if ((strnlen(path, D_TM_MAX_NAME_LEN) +
+			     strnlen(str, D_TM_MAX_NAME_LEN) + 1) <
+			     D_TM_MAX_NAME_LEN) {
+				strncat(path, "/", 1);
+				strncat(path, str, D_TM_MAX_NAME_LEN-1);
+				str = va_arg(args, char *);
+			} else {
+				D_ERROR("Failed to find metric");
+				return -DER_EXCEEDS_PATH_LEN;
+			}
+		}
+		va_end(args);
+		node = d_tm_find_metric(shmemRoot, path);
+		if (metric)
+			*metric = node;
+	}
+
+	if (!node) {
+		rc = d_tm_add_metric(&node, path, D_TM_GAUGE,
+					"N/A", "N/A");
+		if (rc != D_TM_SUCCESS) {
+			D_ERROR("Failed to decrement gauge [%s]  "
+				"Failed to add metric.", path);
+			return rc;
+		}
+		if (metric)
+			*metric = node;
+	}
+
+	if (node->d_tm_type == D_TM_GAUGE) {
+		D_MUTEX_LOCK(&node->lock);
+		node->metric->data.value -= value;
+		D_MUTEX_UNLOCK(&node->lock);
+	} else {
+		D_ERROR("Failed to decrement gauge [%s] on item "
+			"not a gauge.  Operation mismatch.",
+			path);
+		return -DER_OP_NOT_PERMITTED;
+	}
+	return rc;
+}
+
+/*
+ * Finds the node pointing to the given metric described by path name provided
+ */
+d_tm_node_t *
+d_tm_find_metric(uint64_t *cshmemRoot, char *path)
+{
+	char str[256];
+	char *token;
+	char *rest = str;
+	d_tm_node_t *node = NULL;
+	d_tm_node_t *parentNode = d_tm_get_root(cshmemRoot);
+
+	sprintf(str, "%s", path);
+	while ((token = strtok_r(rest, "/", &rest))) {
+		node = d_tm_find_child(cshmemRoot, parentNode, token);
+		if (!node)
+			return NULL;
+		else
+			parentNode =  node;
+	}
+	return node;
+}
+
+/*
+ * Adds a new metric at the specified path, with the given metricType.
+ * An optional short description and long description may be added at this time.
+ * This function may be called by the developer to initialize a metric at init
+ * time in order to avoid the overhead of creating the metric at a more
+ * critical time.
+ */
+int
+d_tm_add_metric(d_tm_node_t **node, char *metric, int metricType,
+		char *shortDesc, char *longDesc)
+{
+	char *str;
+	char *token;
+	char *rest;
+	int buffLen;
+	d_tm_node_t *parentNode;
+	int rc;
+ 	pthread_mutexattr_t mattr;
```

Comment:
(style)  code indent should use tabs where possible
(style)  please, no space before tabs